### DISCUSSION TITLE
Feature Request: Replaceable DummyInput

### DISCUSSION THREAD

**Discussion by dreamwasp** [JedWatson/react-select#4664] [Ideas]
I've run into a situation where I would like to replace the DummyInput rendered when isSearchable is false, as is possible with other components of React-Select.

_- What's your use case?_ 
I need to spread custom props directly onto the HTML `input`, which I also need to maintain its value in the DOM when an element is selected. I am using this with [React-Recurly](https://developers.recurly.com/reference/recurly-js/#elements) which requires both of these things. For instance:
``` html
<form id="my-form">
  <!-- this data-recurly prop + value must be provided -->
  <input type="text" data-recurly="country" value="AU">
  <button>submit</button>
</form>

```

_- What interface do you have in mind?_ 
I would like to be able to replace DummyInput through the `components` prop.

_- Can you point to similar functionality with any existing libraries or components?_
I have a rough draft of when I am proposing pushed on this [branch](https://github.com/JedWatson/react-select/compare/master...dreamwasp:dreamwasp-dummy-input-component), I think it makes sense to have DummyInput be replaceable as Input itself is replaceable.

**Comment by ebonow** [comment]
Greetings @dreamwasp ,

The DummyInput only exists for the purpose of determining is the Select is focused or blurred (which triggers the styling or hiding/showing menu). It has no correlation to any selected value. Additionally, the Input component itself will clear will its value by default unless that behavior is overridden.

What you might be aiming for is not the DummyInput, but rather the generated hidden field which has no official component name but can be seen here: https://github.com/JedWatson/react-select/blob/e6ff154a733d107b637648a4f94cb6634d04ab43/packages/react-select/src/Select.tsx#L1884

This renders only if a name has been provided for the Select so that the form can use this value when submitting a form natively. You should be able to access this onMount using a hook and querying for the hidden input field with the same provided name you gave to your select. Of course you can adjust the query selector to better fit your needs should you have multiple instances.

[Codesandbox demo
](https://codesandbox.io/s/react-select-recurly-hidden-attribute-jkobz)
![Screen Shot 2021-06-28 at 11 58 30 AM](https://user-images.githubusercontent.com/1007579/123689853-8e14d500-d808-11eb-9e76-3dad5bb4ef64.png)

``` jsx
const RecurlySelect = ({ recurly, name, ...props }) => {
  useEffect(() => {
    const query = `input[name="${name}"]`;
    const input = document.querySelector(query);
    input.setAttribute("data-recurly", recurly);
  }, [name, recurly]);

  return <Select {...props} name={name} />;
};
```

Please let me know if this helps or if I am missing something in this implementation.

**Reply by dreamwasp** [reply]
Hi @ebonow - 

This implementation does work for submitting the proper value, but we depend on [cypress-axe](https://github.com/component-driven/cypress-axe) for our accessibility testing and the DummyInput generates an error when used this way: 

```
"Fix any of the following:
  Form element does not have an implicit (wrapped) <label>
  Form element does not have an explicit <label>
  aria-label attribute does not exist or is empty
  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
  Element has no title attribute
  Element has no placeholder attribute
  Element's default semantics were not overridden with role="none" or role="presentation""
```

Alternatively, I suppose the feature ask could be changes to add `role='none'` or `role='presentation'` to the Dummy Input so that it does not trigger this accessibility error.

**Reply by ebonow** [reply]
Excellent that this fulfills your requirement for recurly. 

Regarding accessibility, DummyInput does accept the prop values for  both `aria-label` and `aria-labelledby` if provided to the Select. If you are running cypress-axe, then how is the Input component set up such that it passes but the DummyInput doesn't?

@seancurtis are you aware of any positive/negative impacts to adding roles to the DummyInput?

**Reply by dreamwasp** [reply]
Did some tweaking to how my component was set-up, and finally had success with cypress-axe as long as I match my `aria-labelledby` correctly, thank you so much for your help!
